_posts/2017-12-28-Secure-your-architecture-part2.md

```diff
@@ -0,0 +1,211 @@
+KMS is a fully managed service that lets you manage encryption for your sensitive data on the cloud.
+Most of these KMS's offer the best way for encryption and generate, rotate and destroy your keys. 
+But the KMS is vendor locking so all your keys will stay on the platform.
+To avoid vendor locking, we can implement our own open source version for managing our encryption keys. 
+
+A few examples to get an idea of KMS:
+* [Google Cloud KMS](https://cloud.google.com/kms/){:target="_blank"}
+* [AWS KMS](https://aws.amazon.com/kms/){:target="_blank"}
+* [Azure Vault](https://azure.microsoft.com/en-us/services/key-vault/){:target="_blank"}
+
+
+# Spring Cloud Config Server
+The Spring Cloud Config Server provides a centralized external configuration management backed optionally by a git repository or database.
+Using a REST API for external configuration, Config Server supports encryption and decryption of properties and yml files. 
+First step is installing the Java Cryptography Extension on our local pc.
+>JCE provides a framework and implementation for encryption, key generation, key agreement and message authentication code algorithms. 
+Downloads are available for Java 6, 7 and 8.
+This will allow the config server to use the encryption tool of the JCE.
+
+After the installation, the next step will be securing the config server by adding Spring security to the classpath and configuring your Basic/OAuth2 authentication.
+{% highlight maven %}
+<dependency>
+    <groupId>org.springframework.boot</groupId>
+    <artifactId>spring-boot-starter-security</artifactId>
+</dependency>
+{% endhighlight %}
+
+
+## Key Management
+The config server supports encryption and decryption with a symmetric key or an asymmetric key-pair.
+The choice of which key you will need is within your security terms. 
+The symmetric key is the easiest way to set up but less secure than the asymmetric one. 
+To set up a symmetric key, you just assign a string to the key holder. 
+`encrypt.key=<key>`
+
+To configure these asymmetric keys, we will need a keystore created by the keytool utility from the JDK.
+The public key will encrypt and the private key will decrypt your data.
+
+To create a keystore you can do something like this in your commandline:
+
+{% highlight bash %}
+$> keytool -genkeypair -alias <keyname> -keyalg RSA -keysize 4096 -sigalg SHA512withRSA \
+  -dname "CN=Config Server,OU=JWorks,O=Ordina,L=Mechelen,S=State,C=BE" \
+  -keypass <secret> -keystore config-server.jks -storepass <password>
+{% endhighlight %}
+
+This will generate a keystore for the config server to use. 
+Place it in your repository project and configure it in your yml.
+
+Example yml in the config server:
+
+{% highlight yml %}
+encrypt:
+  key-store:
+    location: classpath:config-server.jks // resource location
+    password: <password> // to unlock the keystore
+    alias: config-server-key // to identify which key in the store is to be used
+    secret: <secret>
+{% endhighlight %}
+
```

Comment:
I would add an alert here that, if you package your keystore within your application jar/war file, the same encryption keys will be used across all of your environments!